pkg/server/server.go

```diff
@@ -0,0 +1,157 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package server provides an implementation of a streaming xDS server
+package server
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	context "golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// ManagementServer is a collection of handlers for streaming discovery requests
+type ManagementServer interface {
+	// Register the handlers in a gRPC server
+	Register(*grpc.Server)
+}
+
+type server struct {
+	cache cache.ConfigCache
+}
+
+func (s *server) Register(srv *grpc.Server) {
+	api.RegisterAggregatedDiscoveryServiceServer(srv, s)
+	api.RegisterEndpointDiscoveryServiceServer(srv, s)
+	api.RegisterClusterDiscoveryServiceServer(srv, s)
+	api.RegisterRouteDiscoveryServiceServer(srv, s)
+	api.RegisterListenerDiscoveryServiceServer(srv, s)
+}
+
+type stream interface {
+	Send(*api.DiscoveryResponse) error
+	Recv() (*api.DiscoveryRequest, error)
+}
+
+func (s *server) handler(stream stream, selector cache.ResourceSelector) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *api.DiscoveryRequest, 0)
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+
+	// the node issuing the last request (assumed to be constant for the session)
+	var node *api.Node
+
+	// unique nonce for req-resp pairs; the server ignores stale nonces
+	var nonce int
+
+	// last successfully applied version as set in the requests
+	var version string
+
+	for {
+		// make a cache request
+		var configCh chan api.DiscoveryResponse
+		if node != nil {
+			configCh = s.cache.Listen(node, selector, version)
+		}
+
+		select {
+		case resp := <-configCh:
+			resp.Nonce = strconv.Itoa(nonce)
+			nonce = nonce + 1
+			if err := stream.Send(&resp); err != nil {
+				return err
+			}
+
+		case req, closed := <-reqCh:
+			// input stream ended or errored out
+			if closed {
+				return nil
+			}
+
+			// ignore stale non-empty nonces
+			if req.GetResponseNonce() != "" && req.GetResponseNonce() != strconv.Itoa(nonce) {
+				continue
+			}
+
+			// update cache request
+			node = req.GetNode()
+			version = req.GetVersionInfo()
+			selector.Names = req.GetResourceNames()
+			if req.TypeUrl != "" {
+				selector.Types = []string{req.TypeUrl}
+			}
+		}
+	}
+}
+
+func (s *server) StreamAggregatedResources(stream api.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	return s.handler(stream, cache.ResourceSelector{})
+}
+
+func (s *server) StreamEndpoints(stream api.EndpointDiscoveryService_StreamEndpointsServer) error {
+	return s.handler(stream, cache.ResourceSelector{
+		Types: []string{cache.EndpointType},
+	})
+}
+
+func (s *server) StreamLoadStats(stream api.EndpointDiscoveryService_StreamLoadStatsServer) error {
+	return errors.New("not implemented")
```

Comment:
You may wish to use `grpc.Errorf(codes.Unimplemented, "not implemented")` I think grpc clients understand this.

Reply:
Thanks, done.